File: ref/algorithmic-trading/apps/app/app/src/main/java/com/trading/app/domain/risk/RiskManager.kt
package com.trading.app.domain.risk

import com.trading.app.domain.model.*

class RiskManager {

    /**
     * 매수 주문 전 리스크 체크
     * 1. 포지션 수 제한
     * 2. 종목 비중 제한
     * 3. 일일 손실 한도
     */
    fun checkBuyOrder(
        signal: Signal,
        riskParams: RiskParams,
        portfolio: PortfolioContext,
        dailyPnlRate: Double = 0.0,
    ): RiskDecision {
        // 1. 포지션 수 제한
        if (portfolio.currentPositionCount >= riskParams.maxPositions) {
            return RiskDecision(
                approved = false,
                action = RiskAction.BLOCK_ORDER,
                reason = "Position limit reached: ${portfolio.currentPositionCount}/${riskParams.maxPositions}",
                eventType = RiskEventType.POSITION_LIMIT,
            )
        }

        // 2. 종목 비중 제한
        val orderAmount = signal.price * signal.quantity
        val weight = orderAmount / portfolio.totalEquity
        if (weight > riskParams.maxPositionWeight) {
            return RiskDecision(
                approved = false,
                action = RiskAction.BLOCK_ORDER,
                reason = "Weight limit exceeded: ${String.format("%.1f", weight * 100)}% > ${String.format("%.1f", riskParams.maxPositionWeight * 100)}%",
                eventType = RiskEventType.WEIGHT_LIMIT,
            )
        }

        // 3. 종목당 최대 금액 제한
        if (orderAmount > riskParams.maxPositionAmount) {
            return RiskDecision(
                approved = false,
                action = RiskAction.BLOCK_ORDER,
                reason = "Amount limit exceeded: ${String.format("%,.0f", orderAmount)}원 > ${String.format("%,.0f", riskParams.maxPositionAmount)}원",
                eventType = RiskEventType.WEIGHT_LIMIT,
            )
        }

        // 4. 자금 부족
        if (orderAmount > portfolio.cash) {
            return RiskDecision(
                approved = false,
                action = RiskAction.BLOCK_ORDER,
                reason = "Insufficient cash: need ${orderAmount.toLong()} but have ${portfolio.cash.toLong()}",
                eventType = RiskEventType.CAPITAL_LIMIT,
            )
        }

        // 5. 일일 손실 한도
        if (dailyPnlRate <= riskParams.dailyLossLimit) {
            return RiskDecision(
                approved = false,
                action = RiskAction.PAUSE_ENGINE,
                reason = "Daily loss limit hit: ${String.format("%.2f", dailyPnlRate * 100)}%",
                eventType = RiskEventType.DAILY_LOSS_LIMIT,
            )
        }

        return RiskDecision(
            approved = true,
            action = RiskAction.APPROVED,
            reason = "OK",
        )
    }

    fun checkSellOrder(signal: Signal): RiskDecision {
        // 매도는 항상 허용 (리스크 관리를 위한 매도는 차단하면 안 됨)
        return RiskDecision(
            approved = true,
            action = RiskAction.APPROVED,
            reason = "Sell always approved",
        )
    }
}

File: ref/algorithmic-trading/apps/engine/src/order/risk-manager.ts
import type Database from 'better-sqlite3';
import type {
  Signal,
  RiskDecision,
  RiskAction,
  RiskEventType,
  RiskParams
} from '@trading/shared/types';
import type { EventBus } from '../event/event-bus.js';
import { EngineEventType } from '../event/event-bus.js';

/**
 * RiskManager: 주문 전 리스크 검증
 * 체인 검증 (첫 거부 시 short-circuit)
 */
export class RiskManager {
  private readonly db: Database.Database;
  private readonly eventBus: EventBus;
  private dailyLossBlocked = false;
  private readonly dailyBlacklist = new Set<string>(); // [SF-3] 종목 당일 블랙리스트

  constructor(deps: { db: Database.Database; eventBus: EventBus }) {
    this.db = deps.db;
    this.eventBus = deps.eventBus;
  }

  /** 매수 주문 리스크 검증 */
  checkBuyOrder(
    signal: Signal,
    riskParams: RiskParams,
    portfolio: {
      cash: number;
      totalEquity: number;
      positionCount: number;
      stockValues: Map<string, number>;
    }
  ): RiskDecision {
    const now = new Date().toISOString();

    // 1. dailyLossBlocked → BLOCK_ORDER
    if (this.dailyLossBlocked) {
      return this.reject(
        'BLOCK_ORDER',
        'DAILY_LOSS_LIMIT',
        'Daily loss limit already triggered'
      );
    }

    // 2. 일일 최대 손실 확인 [MF-7]
    const dailyPnlRate = this.getDailyPnlRate(
      portfolio.totalEquity,
      riskParams.totalCapital
    );
    if (dailyPnlRate <= riskParams.dailyLossLimit) {
      this.dailyLossBlocked = true;
      // RISK_LIMIT_HIT 이벤트 발행 → SessionManager가 PAUSED 전이
      this.eventBus.publish({
        type: EngineEventType.RISK_LIMIT_HIT,
        timestamp: now,
        data: { action: 'PAUSE_ENGINE', type: 'DAILY_LOSS_LIMIT', dailyPnlRate }
      });
      return this.reject(
        'PAUSE_ENGINE',
        'DAILY_LOSS_LIMIT',
        `Daily loss ${(dailyPnlRate * 100).toFixed(2)}% exceeds limit`
      );
    }

    // 3. 최대 보유 종목 수
    if (portfolio.positionCount >= riskParams.maxPositions) {
      return this.reject(
        'REJECTED',
        'POSITION_LIMIT',
        `Position count ${portfolio.positionCount} >= max ${riskParams.maxPositions}`
      );
    }

    // 4. 종목당 최대 비중
    const orderAmount = signal.price * signal.quantity;
    const currentStockValue = portfolio.stockValues.get(signal.stockCode) ?? 0;
    const newWeight = (currentStockValue + orderAmount) / portfolio.totalEquity;
    if (newWeight > riskParams.maxPositionWeight) {
      return this.reject(
        'REJECTED',
        'WEIGHT_LIMIT',
        `Weight ${(newWeight * 100).toFixed(1)}% exceeds max ${(riskParams.maxPositionWeight * 100).toFixed(1)}%`
      );
    }

    // 5. 총 투자금 한도
    if (orderAmount > portfolio.cash) {
      return this.reject(
        'REJECTED',
        'CAPITAL_LIMIT',
        `Order amount ${orderAmount} exceeds cash ${portfolio.cash}`
      );
    }

    // 6. 중복 주문
    if (this.hasActiveOrder(signal.stockCode, 'buy')) {
      return this.reject(
        'REJECTED',
        'ORDER_REJECTED',
        `Duplicate buy order for ${signal.stockCode}`
      );
    }

    // 7. [SF-3] 종목 당일 블랙리스트
    if (this.dailyBlacklist.has(signal.stockCode)) {
      return this.reject(
        'REJECTED',
        'ORDER_REJECTED',
        `${signal.stockCode} is blacklisted today`
      );
    }

    // All pass
    this.eventBus.publish({
      type: EngineEventType.RISK_APPROVED,
      timestamp: now,
      data: { action: 'APPROVED', stockCode: signal.stockCode, side: 'buy' }
    });

    return {
      approved: true,
      action: 'APPROVED',
      reason: 'All risk checks passed'
    };
  }

  /** 매도 주문 리스크 검증 (중복 체크만, 매도는 항상 허용) */
  checkSellOrder(signal: Signal): RiskDecision {
    if (this.hasActiveOrder(signal.stockCode, 'sell')) {
      return this.reject(
        'REJECTED',
        'ORDER_REJECTED',
        `Duplicate sell order for ${signal.stockCode}`
      );
    }

    return {
      approved: true,
      action: 'APPROVED',
      reason: 'Sell order approved'
    };
  }

  /** 매일 초기화 */
  resetDailyState(): void {
    this.dailyLossBlocked = false;
    this.dailyBlacklist.clear();
  }

  /** [SF-3] 종목 블랙리스트 등록 */
  addToBlacklist(stockCode: string): void {
    this.dailyBlacklist.add(stockCode);
  }

  /** 일일 P&L 비율 계산 */
  private getDailyPnlRate(
    currentEquity: number,
    initialCapital: number
  ): number {
    // 간단 계산: (현재 자산 - 초기 자본) / 초기 자본
    // 실제로는 당일 시작 시점 자산 대비 계산해야 하나, 일단 initialCapital 기준
    return (currentEquity - initialCapital) / initialCapital;
  }

  /** 해당 종목의 활성 주문 존재 여부 */
  private hasActiveOrder(stockCode: string, side: string): boolean {
    const row = this.db
      .prepare(
        `
      SELECT COUNT(*) as cnt FROM orders
      WHERE stock_code = ? AND side = ?
      AND status IN ('CREATED', 'SUBMITTED', 'PENDING', 'PARTIAL_FILLED')
    `
      )
      .get(stockCode, side) as { cnt: number } | undefined;
    return (row?.cnt ?? 0) > 0;
  }

  private reject(
    action: RiskAction,
    eventType: RiskEventType,
    reason: string
  ): RiskDecision {
    this.eventBus.publish({
      type: EngineEventType.RISK_REJECTED,
      timestamp: new Date().toISOString(),
      data: { action, eventType, reason }
    });
    return {
      approved: false,
      action,
      reason,
      eventType
    };
  }
}
